### DISCUSSION TITLE
Debian-based Distro Audio Fix when self building

### DISCUSSION THREAD

**Discussion by Konradrundfunk** [AlexandreRouma/SDRPlusPlus#112] [General]
Unfortunatly the  default librtaudio-dev library on the apt is outdated and falsely configured. Therefore you most probably will experience audio problems on linux mint. There is a fix but it requires a bit of manual work for you to do. 

1. You need to clone the repo for a manual install of rtaudio.   `git clone https://github.com/thestk/rtaudio && cd rtaudio `
2. Create a build folder and cd into it. `mkdir build && cd build`
3. Checkout to a working branch `git checkout 5.1.0`
3. Cmake the project without alsa support cause that is what actually breaks it . `cmake -DCMAKE_INSTALL_PREFIX:PATH="/usr" -DRTAUDIO_API_JACK=OFF -DRTAUDIO_API_ALSA=OFF ..`
4. Remove the outdated dependency. `apt remove librtaudio-dev`  **or your rtaudio version**
5. Install the new dependency `sudo make install `
6. Done! Start Sdrpp and enjoy clear sound. 

Thanks to cropinghigh for the cmakeflag.

**Comment by jaymot123** [comment]
That fixed my fluttery audio in Debian Buster too. Thanks!

**Reply by blacksmock** [reply]
I had the same problem with SDR++ v 1.0.1 installed using sdrpp_ubuntu_focal_amd64.deb package on Lubuntu 20.04. This fix cured the problem. Thanks.

**Comment by wa2zkd** [comment]
I am running Linux Mint with Focal ubuntu, runs fine except no audio. Doesn't appear in the pulse vol ctrl system. I installed from a .deb, do I need to build from src?

**Reply by the-gobleo-tinkerer** [reply]
Could you check the Sinks tab and see if you have selected a sink other than "Null"?

**Reply by wa2zkd** [reply]
yes, that was it. Thank You

**Comment by aso824** [comment]
On Linux Mint 20 also `libpulse-dev` need to be installed, and switch `-DRTAUDIO_API_PULSE=ON` to be added to compile it correctly

**Reply by mbott001** [reply]
> On Linux Mint 20 also `libpulse-dev` need to be installed, and switch `-DRTAUDIO_API_PULSE=ON` to be added to compile it correctly

Not being the sharpest crayon in the Linux box, can you list exactly what the entry should be?  More than likely, there will be followup questions later.  :)

**Reply by NicolasGagne** [reply]
Question, do you have acces to the sample rate 11025? 
I don't and want to know is that normal?

**Comment by noman-land** [comment]
Thanks for the help. This fixed my problem. I would also like to point out that there is now a [`5.2.0`](https://github.com/thestk/rtaudio/releases/tag/5.2.0) version of `rtaudio` so step 3 of @Konradrundfunk's instructions can be amended.

**Reply by Konradrundfunk** [reply]
With the latest 5.2 version there seams to be an error so i added the checkout to the earlier version but glad that that helped.

**Reply by noman-land** [reply]
After following your instructions and rebuilding rtaudio, I just re-installed SDR++ and it started working great. I didn't even need to build SDR++ myself.

**Reply by Konradrundfunk** [reply]
Yeah CPP is using some libraries from your system guess that libaudio is one of them so quick fix I guess.

**Comment by gbeyer3** [comment]
@Konradrundfunk 

Thanks for the instructions on building a new librtaudio-dev.  Everything worked as advertised in steps 1-6, (I changed step 3 to 5.2.0) but at step 6 I got an errors, don't know how to resolve.  I searched the code repo for error.h, not found.

sudo make install
[  5%] Building CXX object CMakeFiles/rtaudio.dir/RtAudio.cpp.o
/home/greg/rtaudio/RtAudio.cpp:8525:10: fatal error: pulse/error.h: No such file or directory
 8525 | #include <pulse/error.h>
      |          ^~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [CMakeFiles/rtaudio.dir/build.make:63: CMakeFiles/rtaudio.dir/RtAudio.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:355: CMakeFiles/rtaudio.dir/all] Error 2
make: *** [Makefile:141: all] Error 2

**Reply by Konradrundfunk** [reply]
Do you have pulseaudio installed?

**Comment by gbeyer3** [comment]
@Konradrundfunk  Yes, I do: pulseaudio (1:13.99.1-1ubuntu3.13)

**Comment by gbeyer3** [comment]
I found the post: "With the latest 5.2 version there seams to be an error so i added the checkout to the earlier version but glad that that helped." 

So, I started over again  - deleted contents of /build, then followed all steps again, but checked out 5.1.0.   This time, no error!

But SDR++ is not starting.  I'm going to try re-installing it.  I may have accidentally borked it with a sudo apt autoremove that was suggested when I was doing the steps.   Lesson -- stick to the plan, LOL.

**Reply by HemiBob** [reply]
I did that too and messed it all up :-(

**Comment by gbeyer3** [comment]
Hmm, it wants to install librtaudio-dev . . . hope it does not overwrite the new version I have . . .

**Comment by gbeyer3** [comment]
@Konradrundfunk  rebuilding from the top seems to have resolved the problem.  IMO your audio fix needs to be added to the README, but it's not my project, so I hesitate to horn in.

One thing I noticed reinstalling SDR++ from the .deb file, one of the files that it installs (downloads) is librtaudio-dev.  So the new librtaudio-dev I just built is overwritten by the old, defective one.  Meaning I had to run step 5 & 6 one more time.  I think this will happen every time a new SDR++ is released.  So it seems this issue needs to be fixed.

Again, thanks very much for documenting this fix.

**Reply by jfath** [reply]
Instead of step #5 'apt remove librtaudio-dev', you could run this script which removes the files that will be replaced by 'make install', but prevents the system from trying to re-install librtaudio-dev.

#!/bin/bash

sudo rm /usr/lib/x86_64-linux-gnu/librtaudio.so.6.1.0
sudo rm /usr/lib/x86_64-linux-gnu/librtaudio.so.6
sudo rm /usr/lib/x86_64-linux-gnu/librtaudio.so
sudo rm /usr/include/RtAudio.h
sudo rm /usr/share/rtaudio/RtAudioConfig.cmake
sudo rm /usr/share/rtaudio/RtAudioTargets.cmake
sudo rm /usr/share/rtaudio/RtAudioTargets-noconfig.cmake
sudo rm /usr/lib/x86_64-linux-gnu/pkgconfig/rtaudio.pc

**Reply by AlexandreRouma** [reply]
This is a REALLY bad idea. Don't do this...

**Reply by Konradrundfunk** [reply]
If you already have a utility like apt installed removing libs manually can mess up your system pretty bad like breaking other programs and even the operating system.

**Reply by HemiBob** [reply]
I've only just found this thread after days of frustration, I hope someone can help.
1st, I'm on Mint Xfce so wasn't sure if I needed focal or jammy so I tried focal (nightly build) - it worked but no sound.
So removed that and installed jammy, it worked but sound was intemittant (RSPdx with the API seems to work fine).
I Googled and found this (not here) and ran it (it's missing the git checkout 5.1.0 bit)

"sudo git clone https://github.com/thestk/rtaudio && cd rtaudio
sudo mkdir build && cd build
sudo cmake -DCMAKE_INSTALL_PREFIX:PATH="/usr" -DRTAUDIO_API_JACK=OFF -DRTAUDIO_API_ALSA=OFF ..
sudo apt remove librtaudio-dev
sudo make install"

That removed sdrpp so I re-installed it and went around this loop a few times until now sdrpp no longer loads with the error:
"sdrpp: symbol lookup error: /usr/lib/sdrpp/plugins/audio_source.so: undefined symbol: _ZN7RtAudioC1ENS_3ApiE"

I do have pulseaudio as I get this:
"E: [pulseaudio] pid.c: Daemon already running.
E: [pulseaudio] main.c: pa_pid_file_create() failed."

I'm completely new to Linux, what's the best way forward please? 
I know how to use Synaptic Package Manager to completely remove things like sdrpp but it doesn't show rtaudio. 
I'd like to start from fresh so first things first, focal or jammy?
Thanks Bob

**Reply by AlexandreRouma** [reply]
You forgot (well not actually forget because you noticed the difference but you get what I mean) to `git checkout 5.1.0` after doing `cd rtaudio`. SDR++ Is currently incompatible with rtaudio past version 5.1 due to breaking changes to the library's interface.

**Reply by HemiBob** [reply]
I started again, removed sdrpp and installed focal deb and there appeared an executable sdrpp in the /usr.bin directory.
Then I followed the above steps including 3. Checkout to a working branch git checkout 5.1.0
But when I ran step 5 "sudo apt remove librtaudio-dev" the sdrpp executable disappeared and I was back with
"Failed to execute child process "/usr/bin/sdrpp":Failed to execve:No such file or directory"

Above it says "apt remove librtaudio-dev or your rtaudio version", how do I know which rtaudio version I have, maybe I needed the librtaudio-dev in place?

**Reply by AlexandreRouma** [reply]
Well first reinstalling SDR++ does nothing, so don't waste your time with that. Just install the latest nightly for your OS. You haven't said which *version* of Mint you have (XFCE is just the desktop environment, no impact on the software). When you know what exact version you're using, looking which version of ubuntu it's based on (focal = 20.04, jammy = 22.04, etc).

What they meant by "librtaudio-dev or your rtaudio version" is just to remove whatever rtaudio package you have installed. The name of the package is not always the same so use `apt search rtaudio | grep installed` to figure out what you have.

**Reply by HemiBob** [reply]
Thanks for your patience Alexandra, it would appear I have OS Linux Mint 21.1 Vera - Kernel: Linux 5.15.0-72-generic Architecture: x86-64.
"sudo apt search rtaudio | grep installed" returned nothing.
I did spot that librtaudio6 Ver 5.2.0~ds1-1 is showing as installed in Synaptic PM
Don't know about nightly builds for the OS but I did an update & upgrade.
It's nearly 1:00am here so I'll sleep on it. You've been very helpful, I'm determined to get this working, been on it solid for days now.
Regards, Bob

**Reply by AlexandreRouma** [reply]
> "sudo apt search rtaudio | grep installed" returned nothing.

That's good, means it's not installed yet

>  it would appear I have OS Linux Mint 21.1 Vera
> Don't know about nightly builds for the OS but I did an update & upgrade.

I'm talking about the SDR++ nightly, not the OS. Download the "jammy" deb from here: https://github.com/AlexandreRouma/SDRPlusPlus/releases/tag/nightly

**Reply by HemiBob** [reply]
Yes, it's the same deb I installed before (1.1.0-1018)
It says "Same version is already installed" so I clicked on re-install.
I'm back to where sdrpp starts, controls the RSPdx and is controlled by gpredict but the audio is completely broken into 1/2 second bursts.
wxtoimg obviously can't decode that, do I now run the above 7 steps again, as step 5 breaks sdrpp?
Definitely off to bed now, will resume in the morning.

**Reply by HemiBob** [reply]
I didn't get any further with this yesterday, running step 5 above always removes sdrpp. Since you pointed out that rtaudio is not installed I ran steps 1 to 4 and 6 leaving out step 5
This time sdrpp didn't get removed but the sound is still choppy and unusable.

So, I wondering if I removed sdrpp (with Synaptic PM) and then re-installed but this time running: 

sudo apt install libfftw3-dev libglfw3-dev libglew-dev libvolk2-dev ibsoapysdr-dev libairspyhf-dev libiio-dev libad9361-dev librtaudio-dev
libhackrf-dev zstd
BUT OMITTING "librtaudio-dev" 

Then I wouldn't need to run step 5, just step 1 to 4 and 6 - would that work or is this just the same as I already did?

alternatively, if the problem is with Mint, should I rebuild the computer with a different distribution, and if so what would you suggest? It's a fairly old machine and not very powerful so a lightweight distro like Bodhi would be best maybe?

**Comment by nemanjan00** [comment]
This issue also exists on Arch Linux and I have reported it there

https://gitlab.archlinux.org/archlinux/packaging/packages/rtaudio/-/issues/1

**Comment by MatiasSaibene** [comment]
Here is a custom PKGBUILD (to build rtaudio) for Arch and derivatives.
Just remove the .txt extension to use it with makepkg.
[PKGBUILD.txt](https://github.com/user-attachments/files/16913666/PKGBUILD.txt)

**Comment by BenLand100** [comment]
Hit this issue on gentoo; disabling ALSA support in rtaudio does indeed fix the problem. This is not a misconfiguration or version issue -- the ALSA support in rtaudio simply has limitations and should not be enabled.

**Reply by AlexandreRouma** [reply]
it's a limitation of ALSA itself which does not do adaptive resampling (required since the clock of the SDR isn't synced to the clock of the sound card), not really RtAudio. The problem is why RtAudio defaults to ALSA over any audio server.

**Reply by LexiTree** [reply]
I'm about to 'try the fix' - until now I've just found that toggling on/off/on/off the source will ultimately eventually one time produce good results.  Reading now that "SDR isn't synched to the clock ..." leads me to think the reasons my on/off/on/again sometimes works is that sometimes I just happenstance/luckily line up good enough for the clocks to match ... does that make any sense?  Regardless I'm going to rebuild rtaudio without alsa and (hopefully) hear things working correctly all the time ...

**Reply by LexiTree** [reply]
<img width="1284" height="156" alt="grabber-2025Jul25-231351" src="https://github.com/user-attachments/assets/7dd4b2ba-0a3f-4733-8c65-41f4c2904114" />

Worked like a charm - except for one detail I was surprised by ... now in the Sinks section, under Audio, I only have listed "Built-in Audio Analog Stereo".  I expected to see only PulseAudio listed - so this is the opposite to my believed understanding of things ...

Buuuut , it works! so - I'm only irked at not quite knowing what is going on - even as I'm delighted that the sound output is chop-free!

**Comment by jseyfert3** [comment]
I am running Mint 22.1. The deb package of SDR++ had no audio (no audio output seen in Sinks). I removed it, then I cloned the SDR++ repo and built SDR++ from source. The audio worked.

I did _not_ have to build rtaudio from source. I have librtaudio-dev and librtaudio6 installed via apt, both 5.2.0, exactly the same as were installed as dependencies when I installed SDR++ from source. So building SDR++ from source on my computer is doing something different than the nightly build is doing?

To build I followed the instructions on the main repo page. Clone repo, cd into folder, mkdir build. But instead of installing modules and deps manually, I just did `cmake ..`, saw what library was missing, installed with apt, and repeat until it built. Then `make -j12` and `sudo make install`. Opened the newly built and installed SDR++ and audio worked.

**Comment by wamayumi** [comment]
If you are using arch linux, install alsa-plugins and change the sinks' default alsa device to pulseaudio sound server.
For debian, install the libasound2-plugins package and do the same as above.

**Reply by taeric** [reply]
Do I understand that this would be instead of recompiling the librtaudio code?

**Reply by wamayumi** [reply]
I intended to write that meaning. Sorry for insufficient description.

**Comment by zt64** [comment]
On Fedora 42 I'm using the sdr++ package from https://copr.fedorainfracloud.org/coprs/rezso/SDR/. v5.1.0 was building librtaudio.so.6 but sdr++ expects librtaudio.so.7. Checking out rtaudio v6.0.1 built and seems to work fine so far.

**Comment by wcdiver** [comment]
When you have debian you cannot just simply remove the librtaudio-dev. It will always ask to remove sdrpp. So you have to do it the debian way.


**sudo apt install devscripts build-essential fakeroot dpkg-dev**

**apt source librtaudio-dev**

**sudo apt build-dep  librtaudio-dev**


then cd to the directory rtaudio-$version/debian
edit the **rules** file

Find this section

override_dh_auto_configure:

Change it so it looks like this
override_dh_auto_configure:
	dh_auto_configure -- \\
		--without-jack \\
                --without-alsa \

then build the package with **debuild -us -uc**

*.deb packages should appear in your home folder (or where you've downloaded the sources of the rtaudio)

Simply install them with **sudo dpkg -i**
Now, when there will be update of the librtaudio-dev package it will override this settings. So either you will do it again or hold packages with apt.

**Reply by mcflan** [reply]
This fixed it for me (popOS - based on ubuntu 22.04) both with official release and building from source. Was getting segfaults on start-up in `AudioSink::selectById(int) () from /usr/lib/sdrpp/plugins/audio_sink.so` and messages like `error determining Jack input/output channels!`. 

Thanks for the detailed instructions on doing it the "Debian way". Much appreciated.

**Comment by BriellaBugs** [comment]
Worked great for Arch Linux, the only difference being `yay -R rtaudio`, works the same way

**Comment by riggsd** [comment]
I believe that this PR fixes the RTAudio 6.0 audio underrun problem, at least for systems with PulseAudio compatibility like Debian Trixie's PipeWire. It's essentially a one-line change to use RTAudio's PulseAudio compatibility instead of its apparently-broken ALSA compatibility.

https://github.com/AlexandreRouma/SDRPlusPlus/pull/1728
